pkg/remediator/reconcile/worker_test.go

```diff
@@ -44,84 +44,144 @@ import (
 // TestWorker_Run_Remediates verifies that worker.Run remediates declared
 // objects added to the queue.
 func TestWorker_Run_Remediates(t *testing.T) {
-	ctx := context.Background()
-
-	existingObjs := []client.Object{
-		fake.ClusterRoleBindingObject(syncertest.ManagementEnabled),
-		fake.ClusterRoleObject(syncertest.ManagementEnabled),
-	}
-	declaredObjs := []client.Object{
-		fake.ClusterRoleBindingObject(syncertest.ManagementEnabled),
-		fake.ClusterRoleObject(syncertest.ManagementEnabled),
-	}
-	changedObjs := []client.Object{
-		queue.MarkDeleted(ctx, fake.ClusterRoleBindingObject()),
-		fake.ClusterRoleObject(syncertest.ManagementEnabled,
-			core.Label("new", "label")),
-	}
-	expectedObjs := []client.Object{
-		// CRB delete should be reverted
-		// TODO: Upgrade FakeClient to increment UID after deletion
-		fake.ClusterRoleBindingObject(syncertest.ManagementEnabled,
-			core.UID("1"), core.ResourceVersion("1"), core.Generation(1),
-		),
-		// Role change should be reverted
-		fake.ClusterRoleObject(syncertest.ManagementEnabled,
-			core.UID("1"), core.ResourceVersion("3"), core.Generation(1),
-		),
+	testCases := []struct {
+		name         string
+		existingObjs []client.Object
+		declaredObjs []client.Object
+		changedObjs  []client.Object
+		eventObjs    []client.Object
+		expectedObjs []client.Object
+	}{
+		{
+			name: "revert delete",
+			existingObjs: []client.Object{
+				fake.ClusterRoleBindingObject(syncertest.ManagementEnabled),
+			},
+			declaredObjs: []client.Object{
+				fake.ClusterRoleBindingObject(syncertest.ManagementEnabled),
+			},
+			changedObjs: []client.Object{
+				queue.MarkDeleted(context.Background(), fake.ClusterRoleBindingObject()),
+			},
+			eventObjs: []client.Object{
+				queue.MarkDeleted(context.Background(), fake.ClusterRoleBindingObject()),
+			},
+			expectedObjs: []client.Object{
+				// TODO: Upgrade FakeClient to increment UID after deletion
+				fake.ClusterRoleBindingObject(syncertest.ManagementEnabled,
+					core.UID("1"), core.ResourceVersion("1"), core.Generation(1),
+				),
+			},
+		},
+		{
+			name: "revert watch filter label removal",
+			existingObjs: []client.Object{
+				fake.ClusterRoleBindingObject(syncertest.ManagementEnabled,
+					core.Label("example-label", "example-value")),
+			},
+			declaredObjs: []client.Object{
+				fake.ClusterRoleBindingObject(syncertest.ManagementEnabled,
+					core.Label("example-label", "example-value")),
+			},
+			changedObjs: []client.Object{
+				// Update object to remove label
+				fake.ClusterRoleBindingObject(syncertest.ManagementEnabled),
+			},
+			eventObjs: []client.Object{
+				// Watch server treats update to remove required label as a delete.
```

Comment:
Do you have a link to the code? I'm curious about what labels are required.

Reply:
When you use label selectors in the watch, it sends a delete event when the selected label is removed.

I asked apimachinery if this was a bug and they said it was working as intended. I haven't gone and tracked down the code for it in the apiserver. It's probably buried in the storage layer.

With our current watches, we don't have any label selectors, so this hasn't been an issues. But for https://github.com/GoogleContainerTools/kpt-config-sync/pull/1301 it's causing two of the conflict e2e tests to fail because they not correctly detecting the label removal, so there's no conflict error.

Reply:
More explanation here: go/config-sync-watch-filter#heading=h.tpnna8sp25vg

Reply:
I found a test for similar behavior upstream using a field selector: https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/storage/testing/watcher_tests.go#L103

Reply:
Actually the "filtering" test immediately after that shows the same behavior for label selectors.